aya/src/maps/xdp/cpu_map.rs

```diff
@@ -0,0 +1,178 @@
+//! An array of available CPUs.
+
+use std::{
+    borrow::{Borrow, BorrowMut},
+    os::fd::{AsFd, AsRawFd},
+};
+
+use aya_obj::generated::{bpf_cpumap_val, bpf_cpumap_val__bindgen_ty_1};
+
+use crate::{
+    maps::{check_bounds, check_kv_size, IterableMap, MapData, MapError},
+    programs::ProgramFd,
+    sys::{bpf_map_lookup_elem, bpf_map_update_elem, SyscallError},
+    Pod, FEATURES,
+};
+
+/// An array of available CPUs.
+///
+/// XDP programs can use this map to redirect packets to a target
+/// CPU for processing.
+///
+/// # Minimum kernel version
+///
+/// The minimum kernel version required to use this feature is 4.15.
+///
+/// # Examples
+/// ```no_run
+/// # let elf_bytes = &[];
+/// use aya::maps::xdp::CpuMap;
+///
+/// let mut bpf = aya::BpfLoader::new()
+///     .set_max_entries("CPUS", aya::util::nr_cpus().unwrap() as u32)
+///     .load(elf_bytes)
+///     .unwrap();
+/// let mut cpumap = CpuMap::try_from(bpf.map_mut("CPUS").unwrap())?;
+/// let flags = 0;
+/// let queue_size = 2048;
+/// for i in 0u32..8u32 {
+///     cpumap.set(i, queue_size, None, flags);
+/// }
+///
+/// # Ok::<(), aya::BpfError>(())
+/// ```
+#[doc(alias = "BPF_MAP_TYPE_CPUMAP")]
+pub struct CpuMap<T> {
+    inner: T,
+}
+
+impl<T: Borrow<MapData>> CpuMap<T> {
+    pub(crate) fn new(map: T) -> Result<Self, MapError> {
+        let data = map.borrow();
+
+        if FEATURES.cpumap_prog_id() {
+            check_kv_size::<u32, bpf_cpumap_val>(data)?;
+        } else {
+            check_kv_size::<u32, u32>(data)?;
+        }
+
+        Ok(Self { inner: map })
+    }
+
+    /// Returns the number of elements in the array.
+    ///
+    /// This corresponds to the value of `bpf_map_def::max_entries` on the eBPF side.
+    pub fn len(&self) -> u32 {
+        self.inner.borrow().obj.max_entries()
+    }
+
+    /// Returns the queue size and possible program for a given CPU index.
```

Comment:
What is a "possible" program? Does this mean prog_id=0 if there's no program? In
which case shouldn't we use Option<NonZeroU32>?

Reply:
Indeed. I switched `prog_id` to `Option<NonZero<u32>>`, and reworded to "optional" program.

Reply:
huh indeed. fixed.